doc/designs/quic-design/quic-ackm.md

```diff
@@ -0,0 +1,412 @@
+QUIC ACK Manager
+================
+
+The QUIC ACK manager is responsible for, on the TX side:
+
+  - Handling received ACK frames
+  - Generating notifications that a packet we sent was delivered successfully
+  - Generating notifications that a packet we sent was lost
+
+On the RX side, it is responsible for:
+
+  - Generating ACK frames for later transmission in response to packets we
+    received
+
+In order to allow it to perform these tasks, the ACK manager must:
+
+  - be notified of all transmitted packets
+  - be notified of all received datagrams
+  - be notified of all received packets
+  - be notified of all received ACK frames
+  - be notified when a packet number space is discarded
+  - be notified when its loss detection deadline arrives
+
+The ACK manager consumes:
+
+  - an arbitrary function which returns the current time;
+  - a RTT statistics tracker;
+  - a congestion controller.
+
+The ACK manager provides the following outputs:
+
+  - It indicates the current deadline by which the loss detection
+    event should be invoked.
+
+  - It indicates when probes should be generated.
+
+  - It indicates what ACK frames should be generated.
+
+  - It indicates the largest unacknowledged packet number
+    for a given packet number space.
+
+  - It calls a callback for each transmitted packet it is notified
+    of, specifying whether the packet was successfully acknowledged by the peer,
+    lost or discarded.
+
+  - It may communicate with a congestion controller, causing the
+    congestion controller to update its state.
+
+  - It may communicate with a RTT statistics tracker, causing it
+    to update its state.
+
+In this document, “the user” refers to the system which makes use of the ACK
+manager.
+
+Utility Definitions
+-------------------
+
+There are three QUIC packet number spaces: Initial, Handshake and Application
+Data.
+
+```c
+/* QUIC packet number spaces. */
+#define QUIC_PN_SPACE_INITIAL       0
+#define QUIC_PN_SPACE_HANDSHAKE     1
+#define QUIC_PN_SPACE_APP           2
+#define QUIC_PN_SPACE_NUM           3
+```
+
+Packet numbers are 62-bit values represented herein by `QUIC_PN`.
+`QUIC_PN_INFINITE` evaluates to an invalid QUIC packet number value.
+
+```c
+/* QUIC packet number representation. */
+typedef uint64_t QUIC_PN;
+#define QUIC_PN_INFINITE            UINT64_MAX
+```
+
+Instantiation
+-------------
+
+The QUIC ACK manager is instantiated as follows:
+
+```c
+typedef struct quic_ackm_st QUIC_ACKM;
+
+QUIC_ACKM *QUIC_ACKM_new(OSSL_TIME (*now)(void *arg),
+                         void *now_arg,
+                         QUIC_RTT *rtt,
+                         OSSL_CC_METHOD *cc_method,
+                         OSSL_CC_DATA *cc_data);
+void QUIC_ACKM_delete(QUIC_ACKM *ackm);
+```
+
+The function pointer `now` is invoked by the ACK manager to obtain the current
```

Comment:
This is desired because it makes the functionality far easier to test. The time function can be replaced with a mock which returns a test-controlled timing value.

Reply:
At some point we will probably want some debugging/testing callback from the ossl_time_now() itself.

Reply:
I favour instrumenting or replacing ossl_time_now() for testing.